test/common/upstream/edf_scheduler_test.cc

```diff
@@ -134,6 +175,114 @@ TEST(EdfSchedulerTest, ManyPeekahead) {
+  for (const auto picks : all_picks) {
+    // Populate sched1 one entry after the other.
+    EdfScheduler<double> sched1;
+    for (uint32_t i = 0; i < num_entries; ++i) {
+      sched1.add(*entries[i], entries[i]);
+    }
+    // Perform the picks on sched1.
+    for (uint32_t i = 0; i < picks; ++i) {
+      sched1.pickAndAdd([](const double& w) { return w; });
+    }
+
+    // Create sched2 with pre-built and pre-picked entries.
+    EdfScheduler<double> sched2 = EdfScheduler<double>::createWithPicks(
+        entries, [](const double& w) { return w; }, picks);
+
+    compareEdfSchedulers(sched1, sched2);
+  }
+}
+
+// Emulates first-pick scenarios by creating a scheduler with the given
+// weights and a random number of pre-picks, and validates that the next pick
+// of all the weights is close to the given weights.
+void firstPickTest(const std::vector<double> weights) {
+  TestRandomGenerator rand;
+  ASSERT(std::accumulate(weights.begin(), weights.end(), 0.) == 100.0);
+  // To be able to converge to the expected weights, a decent number of iterations
+  // should be used. If the number of weights is large, the number of iterations
+  // should be larger than 10000.
+  constexpr uint64_t iterations = 1e6;
+  // The expected range of the weights is [0,100). If this is no longer the
+  // case, this value may need to be updated.
+  constexpr double tolerance_pct = 1.0;
+
+  // Set up the entries as simple integers.
+  std::vector<std::shared_ptr<size_t>> entries;
+  entries.reserve(weights.size());
+  for (size_t i = 0; i < weights.size(); ++i) {
+    entries.emplace_back(std::make_shared<size_t>(i));
+  }
+
+  absl::flat_hash_map<size_t, int> sched_picks;
+  auto calc_weight = [&weights](const size_t& i) -> double { return weights[i]; };
+
+  for (uint64_t i = 0; i < iterations; ++i) {
+    // Create a scheduler with the given weights with a random number of
+    // emulated pre-picks.
+    uint32_t r = rand.random();
+    auto sched = EdfScheduler<size_t>::createWithPicks(entries, calc_weight, r);
+
+    // Perform a "first-pick" from that scheduler, and increase the counter for
+    // that entry.
+    sched_picks[*sched.pickAndAdd(calc_weight)]++;
+  }
+
+  // Validate that the observed distribution and expected weights are close.
+  ASSERT_EQ(weights.size(), sched_picks.size());
+  for (const auto& it : sched_picks) {
+    const double expected = calc_weight(it.first);
+    const double observed = 100 * static_cast<double>(it.second) / iterations;
+    EXPECT_NEAR(expected, observed, tolerance_pct);
+  }
+}
+
+// Validates that after creating schedulers using the createWithPicks (with random picks)
+// and then performing a "first-pick", the distribution of the "first-picks" is
+// equal to the weights.
+TEST_F(EdfSchedulerTest, SchedulerWithRandomPicksFirstPickDistribution) {
+  firstPickTest({25.0, 75.0});
+  firstPickTest({1.0, 99.0});
+  firstPickTest({50.0, 50.0});
+  firstPickTest({1.0, 20.0, 79.0});
+}
```

Comment:
FWIW, given how counter-intuitive some of the quirks we've seen in WRR/EDF, even if just offline, I'd suggest taking  two entries and sweeping incrementally at 0.001 from {100, 0} to {0, 100} or something like that to verify this is solved for good :)

Reply:
Seems like a test that will have a longish execution time, but I can try to create it.

Reply:
Yeah. I think just running it once offline would be fine and having a faster version checked in is fine (e.g. every 0.1). Basically, just want to make sure all the corner cases are gone and there aren't any unknown unknowns.

Reply:
Unfortunately it can only run ~80 iterations before timing-out (300s).
I'll run it locally but if there's a better idea on testing it, I'd be happy to implement it.

Reply:
Yeah, maybe just run locally and then check in something with a bunch of points that fits in a reasonable test (just enhancing what is there maybe a bit with extreme points like 0.0001 etc.).

Reply:
While validating found a strange case (~50 of this out of 100K cases) where the math and floating-point arithmetics did not agree:
floor(A*B) / A was greater than B.
Changed the code a bit to adjust for this. Specifically I've capped the number of picks to ~400M, and add a check for the above corner case, that results in with more iterations for the second part of the algorithm, but keeps the correctness of the EDF scheduler initialization.

PTAL!